### ISSUE TITLE
Slow Recursive implementation of Stirling Numbers

### ISSUE THREAD

**Issue by hakobyant** [sympy/sympy#11650]
Hi folks,

I noticed the other day that the current implementation for Stirling Numbers is recursive and thus it works very slowly. I was wondering if there is any idea behind it why it should be recursive.

If there is no specific reason then I would like to work on this to update the implementation to a dynamic one.

**Comment by guoyicong** [comment]
Same question here. What's the reason for the recursive implementation? I wrote a dynamic programming version...

**Comment by sachin-4099** [comment]
Why is this issue open? Is there anything which still needs to be addressed? 
Is Dynamic programming solution needs to be be implemented?
@oscarbenjamin

**Comment by oscarbenjamin** [comment]
I haven't looked at this at all myself. If you are interested then you can check whether the implementation is recursive or slow.

**Comment by sachin-4099** [comment]
How to see whether it is slow or not?
Though I can see the code is indeed a recursive implementation and therefore will definitely be slow for large numbers.

**Comment by oscarbenjamin** [comment]
There is nothing inherently slow about recursion although we will hit the recursion limit of we recurse too many times. Branched recursion can be very slow though - the classic example of this is computing numbers in the fibonacci sequence recursively using `fib(n) = fib(n-1) + fib(n-2)` where the number of calls to `fib` grows exponentially in `n`. That is an example where "dynamic programming" can make things more efficient. As I said though I haven't looked at the code for the Stirling numbers in SymPy so I don't know whether that is what happens.

**Comment by sachin-4099** [comment]
Yes it is exceeding maximum recursion depth.

[stirling(1000, i, kind=1) for i in range(1001)]
RuntimeError: maximum recursion depth exceeded while calling a Python object

**Comment by sachin-4099** [comment]
I have a DP space optimized solution.
Recursive implementation takes exponential time, but DP implementation will take O(n*k) time.
Shall I go ahead and implement it?

**Comment by oscarbenjamin** [comment]
If it is an improvement then go ahead

**Comment by sachin-4099** [comment]
Yeah, it is an improvement, just please review it and merge it after the build.
